### PR TITLE
chore(storage): merge custom client implementation

### PR DESCRIPTION
**Pull request by HuiSF**
# MERGE WITH CREATING A MERGE COMMIT

#### Description of changes

This PR merges the feature branch [`v5/custom-clients`](https://github.com/aws-amplify/amplify-js/tree/v5/custom-clients) into the `main` branch.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/AwsClients/S3/copyObject.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import {
```
Comment (jimblanc) [inline packages/storage/src/AwsClients/S3/copyObject.ts:1]: Slipped through previous reviews, but a handful of these files are missing the copyright headers. 

```
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
```

We really need to get a linter rule setup for this 😮‍💨

**Review by jimblanc — COMMENTED** [review]
Looks to to me! Some things that need to be resolved before before it can be merged, but otherwise looks good

File: packages/storage/src/AwsClients/S3/copyObject.ts (hunk 1 above)
Comment (HuiSF) [reply packages/storage/src/AwsClients/S3/copyObject.ts:1]: 😁 
We could use [addlicense](https://github.com/google/addlicense) to automate this.

**Review by AllanZhengYP — DISMISSED** [review]

**Comment by AllanZhengYP** [comment]
🚀🚀🚀

File: packages/storage/src/AwsClients/S3/copyObject.ts (hunk 1 above)
Comment (jimblanc) [reply packages/storage/src/AwsClients/S3/copyObject.ts:4]: Check my recent PR: https://github.com/aws-amplify/amplify-js/pull/11590 😉

File: CONTRIBUTING.md
Comment (AllanZhengYP) [inline CONTRIBUTING.md:1]: nit: remove the diff

**Review by AllanZhengYP — APPROVED** [review]
🚀🚀🚀 again

**Review by svidgen — APPROVED** [review]

**Pull request merged**